src/lib/optimizer/chunk_statistics/chunk_column_statistics.cpp

```diff
@@ -41,40 +46,43 @@ static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_col
   return build_statistics_from_dictionary(dictionary);
 }
 
-template <typename T>
-static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ValueColumn<T>& column) {
-  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
-  return std::make_shared<ChunkColumnStatistics>();
-}
-
 template <typename T>
 static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(
     const DeprecatedDictionaryColumn<T>& column) {
   const auto& dictionary = *column.dictionary();
   return build_statistics_from_dictionary(dictionary);
 }
 
-template <typename T>
-static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ReferenceColumn& column) {
-  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
-  return std::make_shared<ChunkColumnStatistics>();
-}
-
-template <typename T>
-static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const RunLengthColumn<T>& column) {
-  // TODO(tbjoern): implement build_statistics_from_concrete_column
-  return std::make_shared<ChunkColumnStatistics>();
-}
-
-static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const BaseColumn& column) {
-  return std::make_shared<ChunkColumnStatistics>();
-}
-
 std::shared_ptr<ChunkColumnStatistics> ChunkColumnStatistics::build_statistics(DataType data_type,
                                                                                std::shared_ptr<BaseColumn> column) {
   std::shared_ptr<ChunkColumnStatistics> statistics;
   resolve_data_and_column_type(data_type, *column, [&statistics](auto type, auto& typed_column) {
-    statistics = build_statistics_from_concrete_column(typed_column);
+    using ColumnType = typename std::decay<decltype(typed_column)>::type;
+    using DataType = typename decltype(type)::type;
+
+    if constexpr (std::is_same_v<ColumnType, DictionaryColumn<DataType>> ||
+        std::is_same_v<ColumnType, DeprecatedDictionaryColumn<DataType>>) {
+      // we can use the fact that dictionary columns have an accessor for the dictionary
+      statistics = build_statistics_from_concrete_column(typed_column);
+    } else if constexpr (std::is_base_of_v<BaseEncodedColumn, ColumnType>) {
+      // if we have a generic encoded column we create the dictionary ourselves
+      auto iterable = create_iterable_from_column(typed_column);
+      std::set<DataType> values;
```

Comment:
I don’t think it is that important here, but I think `std::set` might be a bit slow for a million insertions. Perhaps it’s better to use as `std::vector`, reserve a bit of memory, then sort it using `std::sort` and remove duplicate values using `std::unique`, then trim. (see dictionary_encoder.hpp).

Reply:
Wait, The unit of the y-axis is milliseconds. Does that mean higher is worse or better?

Reply:
Worse. That's runtime as far as I remember.